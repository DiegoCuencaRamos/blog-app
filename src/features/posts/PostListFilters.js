import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import { textFilterChanged, sortByFilterChanged } from '../filters/filtersSlice'

export const PostListFilters = ({ filters, textFilterChanged, sortByFilterChanged }) => {

    const onTextChange = (e) => {
        textFilterChanged(e.target.value)
    };

    const onSortByChange = (e) => {
        sortByFilterChanged(e.target.value)
    }

    return (
        <div className="filters">
            <div className="container">
                <div className="filters__wrapper">
                    <div className="filters__content">
                        <input 
                            className="filters__item"
                            type="text"
                            placeholder="Search posts"
                            value={filters.text}
                            onChange={onTextChange}   
                        />
                        <select 
                            className="filters__item"
                            onChange={onSortByChange}
                        >
                            <option value="" defaultValue>Order by</option>
                            <option value="title">Title</option>
                        </select>
                    </div>
                    <Link className="button" to="/create">Add post</Link>
                </div>
                
            </div>
        </div>
    )
}

const mapStateToProps = (state, ownProps) => ({ 
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    textFilterChanged: (text) => dispatch(textFilterChanged(text)),
    sortByFilterChanged: (sortBy) => dispatch(sortByFilterChanged(sortBy)) 
}) 

export default connect(mapStateToProps, mapDispatchToProps)(PostListFilters)

