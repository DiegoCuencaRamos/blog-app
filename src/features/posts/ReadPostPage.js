import React from 'react'
import { connect } from 'react-redux'

export const ReadPostPage = ({ post }) => (
    <div className="read-page">
        <div className="container">
            <h1 className="read-page__title">{post.title}</h1>
            <p className="read-page__text">{post.body}</p>
        </div>
    </div>
)

const mapStateToProps = (state, ownProps) => ({
    post: state.posts.find(post => post.id === ownProps.match.params.id)
})

export default connect(mapStateToProps)(ReadPostPage)