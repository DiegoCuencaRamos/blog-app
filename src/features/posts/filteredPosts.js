const filteredPosts = (posts, { text, sortBy }) => {
    // console.log(text || undefined, sortBy || undefined)
    return posts.filter((post) => {
        return post.title.toLowerCase().includes(text.toLowerCase());
    }).sort((a, b) => {
        if(sortBy === 'title') {
            return a.title > b.title ? 1 : -1;
        }
    })
};

export default filteredPosts;

