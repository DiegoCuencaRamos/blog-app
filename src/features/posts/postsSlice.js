import database from '../../firebase';

const inicialState = []

// Reducer
const postsReducer = (state = inicialState, action) => {
    switch (action.type) {
        case 'posts/postAdded': {
            return [
                ...state,
                {
                    id: action.payload.id,
                    title: action.payload.title,
                    body: action.payload.body
                }
            ]
        }
        case 'posts/postEdited': {
            return state.map(post => {
                if(post.id !== action.payload.id) {
                    return post 
                }

                return {
                    ...post,
                    ...action.payload.updates
                }
            })
        }
        case 'posts/postRemoved': {
            return state.filter(post => post.id !== action.payload)
        }
        case 'posts/postsSet': {
            return action.payload
        }     
        default: 
            return state
    }
}

// Actions creators

// 1. ADD_POST
const postAdded = (post) => ({
    type: 'posts/postAdded',
    payload: {
        ...post
    }
})

const startPostAdded = (postData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            title = '',
            body = ''
        } = postData
        const post = { title, body }

        return database.ref(`users/${uid}/posts`).push(post)
            .then((ref) => {
                dispatch(postAdded({
                    id: ref.key,
                    ...post
                }))
            })

    }
} 

// 2. EDIT_POST
const postEdited = (id, updates) => ({
    type: 'posts/postEdited',
    payload: {
        id,
        updates
    }
})

const startPostEdited = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/posts/${id}`).update(updates)
            .then(() => {
                dispatch(postEdited(id, updates))
            })
    }
}

// 3. REMOVE_POST
const postRemoved = (id) => ({
    type: 'posts/postRemoved',
    payload: id
})

const startPostRemoved = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/posts/${id}`).remove()
            .then(() => {
                dispatch(postRemoved(id))
            })
    }
}

// 4. SET_POSTS
const postsSet = (posts) => ({
    type: 'posts/postsSet',
    payload: posts
})

const startPostsSet = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/posts`).once('value')
            .then((dataSnapshot) => {
                const posts = []

                dataSnapshot.forEach((childDataSnapshot) => {
                    posts.push({
                        id: childDataSnapshot.key,
                        ...childDataSnapshot.val()
                    })
                })

                dispatch(postsSet(posts))
            })
    }
} 

export {
    postAdded,
    startPostAdded,
    postEdited,
    startPostEdited,
    postRemoved,
    startPostRemoved,
    postsSet,
    startPostsSet
}

export default postsReducer