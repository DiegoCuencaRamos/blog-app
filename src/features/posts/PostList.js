import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import filteredPosts from './filteredPosts'

export const PostList = ({ posts, filters }) => {
    
    const selectedPosts = filteredPosts(posts, filters)

    return (
        <div className="list">
            <div className="container">
                {
                    selectedPosts.length === 0
                        ? (
                            <div>
                                <span>No posts to show</span>
                            </div>
                        ) : (
                            <div className="list__table">
                                <div className="list__header">
                                    <div className="list__header-item">
                                        <p>Title</p>
                                    </div>
                                    <div className="list__header-item">
                                        <p>Date</p>
                                    </div>
                                    </div>
                                <div className="list__body">
                                    {
                                        selectedPosts.map(({ id, title }) => (
                                            <Link 
                                                className="list__body-item"
                                                key={id} to={`/edit/${id}`}
                                            >
                                                <p>
                                                    {title}
                                                </p> 
                                            </Link>
                                        ))
                                    }
                                </div>
                            </div>
                        )
                }
            </div>
        </div>
    )
}

const mapStateToProps = (state, ownProps) => ({
    posts: state.posts,
    filters: state.filters
})

export default connect(mapStateToProps)(PostList)