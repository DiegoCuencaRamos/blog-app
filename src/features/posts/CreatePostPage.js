import React from 'react'
import { connect } from 'react-redux'

import PageTitle from '../PageTitle'
import { startPostAdded } from './postsSlice'

export class CreatePostPage extends React.Component {
    constructor(props) {
        super(props)
        // Bind methods
        this.onTitleChange = this.onTitleChange.bind(this)
        this.onBodyChange = this.onBodyChange.bind(this)
        this.onSubmit = this.onSubmit.bind(this)
        //State
        this.state = {
            title: '',
            body: ''
        }
    }

    onTitleChange(e) {
        const title = e.target.value
        this.setState(() => ({ title }))
    }

    onBodyChange(e) {
        const body = e.target.value
        this.setState(() => ({ body }))
    }

    onSubmit(e) {
        e.preventDefault()
        this.props.startPostAdded(this.state)
        this.props.history.push('/')
    }

    render() {
        return (
            <React.Fragment>
                <PageTitle 
                    title={'Add post'} 
                    description={'Add new posts to to your site'} 
                />
                
                <div className="container">
                    <form 
                        className="form"
                        onSubmit={this.onSubmit}
                    >
                        <div className="form__item-wrapper--title">
                            <input 
                                className="form__item--normal"
                                type="text" 
                                value={this.state.title} 
                                placeholder="Post Title" 
                                onChange={this.onTitleChange} 
                            />
                        </div>
                        <div className="form__item-wrapper--normal">
                            <textarea 
                                className="form__item--textarea"
                                value={this.state.body} 
                                placeholder="Post Body" 
                                onChange={this.onBodyChange} 
                            />
                        </div>
                        <div>
                            <button className="button">
                                Save Post
                            </button>
                        </div>
                    </form>
                </div>
            </React.Fragment>
        )
    } 
}

const mapDispatchToProps = (dispatch) => ({
    startPostAdded: (post) => dispatch(startPostAdded(post))
})

export default connect(null, mapDispatchToProps)(CreatePostPage);