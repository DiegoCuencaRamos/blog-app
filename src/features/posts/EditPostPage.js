import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

import PageTitle from '../PageTitle'
import { startPostEdited, startPostRemoved } from './postsSlice'

export class EditPostPage extends React.Component {
    constructor(props) {
        super(props);
        //Bind methods
        this.onTitleChange = this.onTitleChange.bind(this);
        this.onBodyChange = this.onBodyChange.bind(this);
        this.onSaveButtonClicked = this.onSaveButtonClicked.bind(this);
        this.onRemoveButtonClicked = this.onRemoveButtonClicked.bind(this);
        // State
        this.state = {
            title: this.props.post.title,
            body: this.props.post.body
        }
    }

    onTitleChange(e) {
        const title = e.target.value
        this.setState((state, props) => ({ title }))
    }

    onBodyChange(e) {
        const body = e.target.value
        this.setState((state, props) => ({ body }))
    }

    onSaveButtonClicked(e) {
        e.preventDefault()
        this.props.startPostEdited(this.props.post.id, this.state)
        this.props.history.push('/')
    }

    onRemoveButtonClicked() {
        this.props.startPostRemoved(this.props.post.id, this.state)
        this.props.history.push('/')
    }

    render() {
        return (
            <React.Fragment>
                <PageTitle 
                    title={'Edit post'}
                    description={'Make some changes and update your posts'}
                />

                <div className="container">
                    <Link 
                        className="link"
                        to={`/read/${this.props.post.id}`}
                    >
                        > Read
                    </Link>
                    <form className="form">
                        <div className="form__item-wrapper--title">
                            <input 
                                className="form__item--normal"
                                type="text" 
                                value={this.state.title} 
                                onChange={this.onTitleChange}
                            />
                        </div>
                        <div className="form__item-wrapper--normal">
                            <textarea 
                                className="form__item--textarea"
                                value={this.state.body} 
                                onChange={this.onBodyChange}
                            />
                        </div>
                        <div className="form__buttons-wrapper">
                            <button 
                                className="button"
                                onClick={this.onSaveButtonClicked}
                            >
                                Save Post
                            </button>
                            <button 
                                className="button--remove" 
                                onClick={this.onRemoveButtonClicked}
                            >
                                Remove Post
                            </button>
                        </div>
                    </form>
                    
                </div> 
            </React.Fragment>
        )
    }
};

const mapStateToProps = (state, ownProps) => ({
    post: state.posts.find(post => post.id === ownProps.match.params.id)
})


const mapDispatchToProps = (dispatch) => ({
    startPostEdited: (id, updates) => dispatch(startPostEdited(id, updates)),
    startPostRemoved: (id) => dispatch(startPostRemoved(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);