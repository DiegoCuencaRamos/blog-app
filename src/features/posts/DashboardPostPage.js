import React from 'react'

import PageTitle from '../PageTitle'
import PostListFilters from './PostListFilters'
import PostList from './PostList'

export const DashboardPage = (props) => (
    <section>
        <PageTitle 
            title={'Dashboard'} 
            description={'From here you will be able visualice all your posts'}
        />
        <PostListFilters/>
        <PostList />
    </section>
)

export default DashboardPage