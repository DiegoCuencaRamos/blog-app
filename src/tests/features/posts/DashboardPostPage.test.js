import React from 'react'
import { shallow } from 'enzyme'
import { DashboardPage } from '../../../features/posts/DashboardPostPage'

test('should render PostDashboardPage correctly', () => {
    const wrapper = shallow(<DashboardPage />)
    expect(wrapper).toMatchSnapshot()
})