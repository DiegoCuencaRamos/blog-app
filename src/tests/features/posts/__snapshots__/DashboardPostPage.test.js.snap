// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostDashboardPage correctly 1`] = `
<section>
  <PageTitle
    description="From here you will be able visualice all your posts"
    title="Dashboard"
  />
  <Connect(PostListFilters) />
  <Connect(PostList) />
</section>
`;
